src/main/java/com/google/sps/servlets/RoutesStoring.java

```diff
@@ -0,0 +1,107 @@
+import java.io.IOException;
+import java.util.stream.Collectors;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that process information about user's markers. */
+@SuppressWarnings("serial")
+@WebServlet("/storeRoute")
+public class RoutesStoring extends HttpServlet {
+  static class RouteData {
+    String routeName;
+    Marker[] markersData;
+
+    RouteData(String routeName, Marker[] markersData) {
+      this.routeName = routeName;
+      this.markersData = markersData;
+    }
+
+    String getRouteName() {
+      return routeName;
+    }
+
+    Marker[] getMarkersData() {
+      return markersData;
+    }
+  }
+
+  String json;
+
+  /** Processes POST request by storing routes. */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    // Check if user is logged in.
+    if (!userService.isUserLoggedIn()) {
+      response.setContentType("text/html;");
+      response.getWriter().println("<p>ERROR: You are not logged in</p>");
+      response.getWriter().println("<a href=\"index.html\">Go to home page</button>");
+      return;
+    }
+
+    // Read request body.
+    String requestBody =
+        request.getReader().lines().collect(Collectors.joining(System.lineSeparator()));
+
+    // Convert it to Gson Object.
+    Gson gson = new Gson();
+    RouteData gsonObject = gson.fromJson(requestBody, RouteData.class);
+    Marker[] routeMarkers = gsonObject.getMarkersData();
+    String routeName = gsonObject.getRouteName();
+
+    Key userKey = KeyFactory.createKey("User", userService.getCurrentUser().getUserId());
+
+    try {
+      Entity userEntity = datastore.get(userKey);
+      // Create new route entity and make it child of the user.
+      Entity routeEntity = new Entity("Route", userEntity.getKey());
+      routeEntity.setProperty("name", routeName);
+      datastore.put(routeEntity);
+
+      for (Marker marker : routeMarkers) {
+        // Create entities for markers and make them children of the route.
+        Entity markerEntity = new Entity("Marker", routeEntity.getKey());
+
+        markerEntity.setProperty("lat", marker.getLat());
+        markerEntity.setProperty("lng", marker.getLng());
+        markerEntity.setProperty("visitHour", marker.getVisitHour());
+        markerEntity.setProperty("visitMinute", marker.getVisitMinute());
+        markerEntity.setProperty("leaveHour", marker.getLeaveHour());
+        markerEntity.setProperty("leaveMinute", marker.getLeaveMinute());
+        markerEntity.setProperty("markerName", marker.getMarkerName());
+
+        // Store new marker.
+        datastore.put(markerEntity);
+      }
+      response.sendRedirect("index.html");
```

Comment:
This should probably not redirect, seeing that it is called by a javascript fetch rather than a form submit. It would be better to send a status message as JSON, which can be handled by the javascript. This can be done as part of a follow up PR, just add a TODO linking to an issue.